lib/dynamodb-integration.js

```diff
@@ -1,121 +1,119 @@
-          const extraValues = {};
-          additionalAttributes.forEach(([key, dynamoValueType]) => {
-            if (!isUndefined(rawMessage[key])) {
-              extraValues[key] = {
-                [dynamoValueType]: rawMessage[key].toString(),
-              };
-            }
-          });
-
-          return {
-            PutRequest: {
-              Item: {
-                ...extraValues,
-                id: {
-                  S: streamName,
-                },
-                timestamp: {
-                  N: event.timestamp.toString(),
-                },
-                message: {
-                  S: event.message,
-                },
+  return logEvents.splice(0, entryIndex);
+};
+
+const buildPayload = (tableName, streamName, events, additionalAttributesSchema) => {
+  var additionalAttributes = additionalAttributesSchema ? Object.entries(additionalAttributesSchema) : [];
+  return {
+    RequestItems: {
+      [tableName]: events.map(event => {
+        const rawMessage = event.rawMessage;
+        const extraValues = {};
+        additionalAttributes.forEach(([key, dynamoValueType]) => {
+          if (!isUndefined(rawMessage[key])) {
+            extraValues[key] = {
+              [dynamoValueType]: rawMessage[key].toString()
+            };
+          }
+        });
+
+        return {
+          PutRequest: {
+            Item: {
+              ...extraValues,
+              id: {
+                S: streamName
+              },
+              timestamp: {
+                N: event.timestamp.toString()
               },
-            },
-          };
-        }),
-      },
-    };
+              message: {
+                S: event.message
+              }
+            }
+          }
+        };
+      })
+    }
   };
+};
 
-const hasUnprocessedItems = (data) => !isEmpty(data.UnprocessedItems);
+const hasUnprocessedItems = data => !isEmpty(data?.UnprocessedItems);
 
 const isDynamoDbError = (err, data) => err || hasUnprocessedItems(data);
 
-const sendLogsToDynamoDb = (dynamodbClient, payload, cb, times = 3) => new Promise((resolve) => {
-  dynamodbClient.send(new BatchWriteItemCommand(payload), function (err, data) {
-    debug('sent to aws, err: ', err, ' data: ', data)
-    if (isDynamoDbError(err, data) && times > 0) {
-      debug('error during batchWriteItem', err, true)
+const sendLogsToDynamoDb = (dynamodbClient, payload, cb, attempt = 1) =>
+  new Promise(resolve => {
+    dynamodbClient.send(new BatchWriteItemCommand(payload), async function (err, data) {
+      debug('sent to aws, err: ', err, ' data: ', data);
+
+      if (isDynamoDbError(err, data)) {
+        debug('error during batchWriteItem', err, true);
+        if (attempt > LIMITS.MAX_BATCH_WRITING_RETRIES) {
+          debug('Got to 3 attempts, not retrying.');
+          cb();
+          resolve();
+        }
+
+        await new Promise(resolve => setTimeout(() => resolve(), 500 * Math.pow(2, attempt - 1)));
```

Comment:
nip
extract to a const as it hard to read

Reply:
46c9c7b2a41d09abb2936692f74179e9917654e4